dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionPreprocessor.java

```diff
@@ -0,0 +1,205 @@
+     */
+    private Predictor clonePredictor( Predictor predictor, PredictorPreprocessorExpression preEx,
+        DataElement dataElement )
+    {
+        Predictor clone = predictor.toBuilder().build();
+
+        clone.setOutput( dataElement );
+
+        String mainExpression = preEx.getMain();
+        String variable = preEx.getVariable();
+        String clonedExpression = mainExpression.replace( variable, dataElement.getUid() );
+        Expression clonedGenerator = new Expression( clonedExpression, clone.getGenerator().getDescription(),
+            clone.getGenerator().getMissingValueStrategy() );
+        clone.setGenerator( clonedGenerator );
+
+        return clone;
+    }
+
+    /**
+     * Gets the description for a predictor.
+     */
+    private String describe( String expression )
+    {
+        PredictorPreprocessorExpression preEx = new PredictorPreprocessorExpression( expression );
+
+        // Get prefix description (resolve data element group)
+        DataElementGroup deg = getDeg( preEx.getDegUid(), expression );
+        String prefixDescription = preEx.getPrefix().replace( preEx.getTaggedDegUid(), deg.getName() );
+
+        // Find item descriptions for the main expression with a dummy
+        // expression resolving the data element variable to some DE in the DEG
+        DataElement de = getOneDataElement( deg, expression );
+        String mainExpressionWithOneDataElement = preEx.getMain().replace( preEx.getVariable(), de.getUid() );
+        Map<String, String> itemDescriptions = expressionService.getExpressionItemDescriptions(
+            mainExpressionWithOneDataElement, PREDICTOR_EXPRESSION );
+
+        // Get the main expression description
+        String mainDescription = preEx.getMain();
+        for ( Map.Entry<String, String> entry : itemDescriptions.entrySet() )
+        {
+            mainDescription = mainDescription.replace( entry.getKey(), entry.getValue() );
+        }
+
+        // Return the complete description
+        return prefixDescription + PREPROCESSOR_SEPARATOR + mainDescription;
+    }
+
+    /**
+     * Gets the data element group for preprocessing.
+     */
+    private DataElementGroup getDeg( String deUid, String expression )
+    {
+        DataElementGroup deg = idObjectManager.get( DataElementGroup.class, deUid );
+
+        if ( deg == null )
+        {
+            throw new ParserException(
+                format( "Can't find data element group %s in %s", deUid, expression ) );
+        }
+        return deg;
+    }
+
+    /**
+     * Gets one data element from the data element group.
+     */
+    private DataElement getOneDataElement( DataElementGroup deg, String expression )
+    {
+        Set<DataElement> dataElements = deg.getMembers();
+
+        if ( dataElements.isEmpty() )
+        {
+            throw new ParserException(
+                format( "DataElementGroup '%s' is empty when evaluating '%s'", deg.getName(), expression ) );
```

Comment:
This I would use `org.hisp.dhis.feedback.ConflictException`. Again assuming this should end up as the response to the client. A "parse exception" should only be used to indicate that the syntax was wrong.

Reply:
Also changed to `IllegalStateException` for now, and also hope we can revisit all of the exception generation when we integrate the PEG parser.